primo/utils/clustering_utils.py

```diff
@@ -145,3 +177,106 @@ def perform_clustering(wd: WellData, distance_threshold: float = 10.0):
     wd.col_names.register_new_columns({"cluster": "Clusters"})
 
     return clustered_data.n_clusters_
+
+
+def perform_louvain_clustering(
+    wd: WellData, distance_threshold: float = 10.0, cluster_threshold: int = 300
+) -> dict:
+    """
+    Partitions the data into smaller clusters using the Louvain community detection method,
+    limiting each well to connect only with its 10 closest neighbors within a threshold distance.
```

Comment:
Can we make this factor (10 closest neighbors) a parameter?

Reply:
Also a good idea. This is how we directly control the memory requirements of the algorithm (because this gives an upper bound on the number of edges in the graph)

Reply:
Added nearest neighbors as a parameter

Reply:
Added nearest neighbor as a model option^^^